### ISSUE TITLE
Cannot annotate a const generic function with a contract

### ISSUE THREAD

**Issue by zhassan-aws** [model-checking/kani#3667]
This issue was reported in https://github.com/model-checking/verify-rust-std/discussions/143.

I tried this code:

```rust
#[kani::requires(true)]
const fn bar<T>(_v: T)
where T: Sized
{
}
```

using the following command line invocation:

```
kani s.rs -Zfunction-contracts
```

with Kani version: 1fe80f9c5b2

```
$ kani s.rs -Zfunction-contracts
Kani Rust Verifier 0.56.0 (standalone)
error[E0493]: destructor of `T` cannot be evaluated at compile-time
 --> s.rs:2:17
  |
1 | #[kani::requires(true)]
  |                       - value is dropped here
2 | const fn bar<T>(_v: T)
  |                 ^^ the destructor for this type cannot be evaluated in constant functions

error: aborting due to 1 previous error

For more information about this error, try `rustc --explain E0493`.
error: /home/ubuntu/git/kani/target/kani/bin/kani-compiler exited with status exit status: 101
```

**Comment by celinval** [comment]
This is not a Kani issue though. If you remove the contract and try to run it with rustc, the same error occurs:
```rust
// dummy.rs
pub const fn bar<T>(_v: T)
where T: Sized
{
}
```

Running:
```
rustc dummy.rs
```
Returns

<pre><font color="#F66151"><b>error[E0493]</b></font><b>: destructor of `T` cannot be evaluated at compile-time</b>
 <font color="#2A7BDE"><b>--&gt; </b></font>dummy.rs:1:21
  <font color="#2A7BDE"><b>|</b></font>
<font color="#2A7BDE"><b>1</b></font> <font color="#2A7BDE"><b>|</b></font> pub const fn bar&lt;T&gt;(_v: T)
  <font color="#2A7BDE"><b>|</b></font>                     <font color="#F66151"><b>^^</b></font> <font color="#F66151"><b>the destructor for this type cannot be evaluated in constant functions</b></font>
<font color="#2A7BDE"><b>...</b></font>
<font color="#2A7BDE"><b>4</b></font> <font color="#2A7BDE"><b>|</b></font> }
  <font color="#2A7BDE"><b>|</b></font> <font color="#2A7BDE"><b>-</b></font> <font color="#2A7BDE"><b>value is dropped here</b></font>

<font color="#F66151"><b>error</b></font><b>: aborting due to 1 previous error</b></pre>

**Comment by celinval** [comment]
The problem here is that the function has a local parameter that may implement `Drop`, and that will be marked as dead by the end of the functions. But constant functions cannot call drop.

Note that if the function does not drop the value, that is OK:

```rust
pub struct WithDrop;
pub struct Wrap<T>(T);

impl Drop for WithDrop {
    fn drop(&mut self) {}
}

pub const fn will_not_drop(v: WithDrop) -> Wrap<WithDrop>
{
    Wrap(v)
}

pub const fn will_drop(_v: WithDrop)
{
}

```
Note that the function `will_not_drop` is OK, but the `will_drop` has a compilation error:

<pre><font color="#F66151"><b>error[E0493]</b></font><b>: destructor of `WithDrop` cannot be evaluated at compile-time</b>
  <font color="#2A7BDE"><b>--&gt; </b></font>dummy.rs:13:24
   <font color="#2A7BDE"><b>|</b></font>
<font color="#2A7BDE"><b>13</b></font> <font color="#2A7BDE"><b>|</b></font> pub const fn will_drop(_v: WithDrop)
   <font color="#2A7BDE"><b>|</b></font>                        <font color="#F66151"><b>^^</b></font> <font color="#F66151"><b>the destructor for this type cannot be evaluated in constant functions</b></font>
<font color="#2A7BDE"><b>14</b></font> <font color="#2A7BDE"><b>|</b></font> {
<font color="#2A7BDE"><b>15</b></font> <font color="#2A7BDE"><b>|</b></font> }
   <font color="#2A7BDE"><b>|</b></font> <font color="#2A7BDE"><b>-</b></font> <font color="#2A7BDE"><b>value is dropped here</b></font>

<font color="#F66151"><b>error</b></font><b>: aborting due to 1 previous error</b>

<b>For more information about this error, try `rustc --explain E0493`.</b></pre>

I'm going to close this for now, but please let me know if you have another example in mind.

**Comment by zhassan-aws** [comment]
@celinval Thanks for clarifying. In the original issue (https://github.com/model-checking/verify-rust-std/discussions/143), the error only appears when the function is annotated with a requires clause.

**Comment by zhassan-aws** [comment]
Here is a small example of the issue in https://github.com/model-checking/verify-rust-std/discussions/143:
```rust
#![feature(const_ptr_write)]

struct Foo<T> {
    ptr: *const T,
}

impl<T: Sized> Foo<T> {
    #[cfg_attr(kani, kani::requires(true))]
    const unsafe fn bar(self, v: T)
    where
        T: Sized,
    {
        unsafe { core::ptr::write(self.ptr as *mut T, v) };
    }

}

fn main() {}
```
rustc succeeds:
```bash
$ rustc w.rs 
warning: struct `Foo` is never constructed
 --> w.rs:3:8
  |
3 | struct Foo<T> {
  |        ^^^
  |
  = note: `#[warn(dead_code)]` on by default

warning: method `bar` is never used
 --> w.rs:9:21
  |
7 | impl<T: Sized> Foo<T> {
  | --------------------- method in this implementation
8 |     #[cfg_attr(kani, kani::requires(true))]
9 |     const unsafe fn bar(self, v: T)
  |                     ^^^

warning: 2 warnings emitted
```
but Kani fails:
```bash
$ kani w.rs 
Kani Rust Verifier 0.56.0 (standalone)
warning: the feature `const_ptr_write` has been stable since 1.83.0 and no longer requires an attribute to enable
 --> w.rs:1:12
  |
1 | #![feature(const_ptr_write)]
  |            ^^^^^^^^^^^^^^^
  |
  = note: `#[warn(stable_features)]` on by default

error[E0493]: destructor of `T` cannot be evaluated at compile-time
 --> w.rs:9:31
  |
8 |     #[cfg_attr(kani, kani::requires(true))]
  |                                         - value is dropped here
9 |     const unsafe fn bar(self, v: T)
  |                               ^ the destructor for this type cannot be evaluated in constant functions
```

**Comment by celinval** [comment]
Thank you for the new example @zhassan-aws. I was able to reproduce it on my side. The problem here seems to be coming from the logic Kani encodes to stub the function.

During stubbing, Kani assign a new value to the return value instead of using the value passed as an argument like the other modes. Thus, the function could potentially drop the value, and the compilation fails.

One potential fix that comes to mind would be to change Kani to do a dummy assignment of the input variables to local variables inside the generated closure. Since the closure doesn't get called in the constant context, the compiler will no longer complain about the drop.
